### PR TITLE
[MRG] Move creation of result object to `Optimizer` class

### LINKED ISSUE #315: Optimizer does not create `OptimizeResult` objects
Ask and tell interface does not create a `OptimizeResult` object like `XXX_minimize` does which makes it harder to use the plotting tools in `skopt`.

`Optimizer.tell()` could return a result object that is compatible with our plotting tools and for API uniformity reasons.

---

This was pointed out in #305

### PR DESCRIPTION
**Pull request by betatim**
Fixes #315

Move the call to `create_result` to `Optimizer` so that we can use the plot tools we have with ask-and-tell.

### CODE REVIEW

File: skopt/space/space.py [hunk 1]
```diff
@@ -214,6 +214,9 @@ def inverse_transform(self, Xt):
     def bounds(self):
         return (self.low, self.high)
 
+    def __contains__(self, point):
+        return self.low <= point <= self.high
```
Comment (betatim) [inline skopt/space/space.py]: are these the right comparison operators? (We should document this in the doc string)

File: skopt/space/space.py [hunk 2]
```diff
@@ -284,6 +287,9 @@ def inverse_transform(self, Xt):
...
```
Comment (betatim) [inline skopt/space/space.py]: are these the right comparison operators? (We should document this in the doc string)

**Comment by betatim** [comment]
As well as moving the creation of the results object I also added some bounds checking in this PR. Ready for reviewing.

File: skopt/space/space.py (hunk 2 above)
Comment (glouppe) [inline skopt/space/space.py]: We have documented the upper bound of a `Real` dimension to be exclusive.

**Comment by glouppe** [comment]
Looks good to me, except the comment for the upper bound.

File: skopt/space/space.py (hunk 2 above)
Comment (betatim) [inline skopt/space/space.py]: Sorry, didn't spot it in the doc string :(

**Comment by betatim** [comment]
Fixed the operators in the bounds checking and modified the tests a bit. They were using identical comparison operators for both `Integer` and `Real`. However one is inclusive and the other exclusive at the upper boundary.

**Comment by betatim** [comment]
Mistake in my implementation somewhere or is our docstring wrong? https://travis-ci.org/scikit-optimize/scikit-optimize/jobs/203851498

**Comment by betatim** [comment]
I think it is because [`fmin_l_bfgs_b()`](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.optimize.fmin_l_bfgs_b.html) interprets the bounds differently (both ends inclusive) to us.

Ahh yiis: https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb/README

File: skopt/space/space.py [hunk 3]
```diff
@@ -214,15 +214,18 @@ def inverse_transform(self, Xt):
...
         return (self.low, self.high)
 
+    def __contains__(self, point):
+        return self.low <= point < self.high
+
     @property
     def transformed_bounds(self):
         if self.transform_ == "normalize":
-            return 0.0, 1.0
+            return 0.0, 1.0 - np.finfo(float).eps
```
Comment (betatim) [inline skopt/space/space.py]: Is this the best we can do? Or should we modify the upper edge just before the call to BFGS?

File: skopt/space/space.py [hunk 4]
```diff
@@ -214,15 +214,19 @@ def inverse_transform(self, Xt):
...
 
+    def __contains__(self, point):
+        return self.low <= point < self.high
+
     @property
     def transformed_bounds(self):
+        eps = np.finfo(float).eps
         if self.transform_ == "normalize":
-            return 0.0, 1.0
+            return 0.0, 1.0 - eps
```
Comment (betatim) [inline skopt/space/space.py]: Is this the best we can do? Or should we modify the upper edge just before the call to BFGS?

**Comment by betatim** [comment]
Decided to change the behaviour of `Real` to be inclusive of upper edge to be consistent with BFGS behaviour. Not sure if this adaptation of `scipy.stats.distributions.uniform` does what I think it does.

**Comment by betatim** [comment]
`np.all(p in self.space for p in x)` can you spot the bug in this?

Time to merge this.

**Comment by glouppe** [comment]
Looks good to me! Didnt know about `nextafter`, learned something :)

**Pull request merged**